File: src/main/java/io/jsql/sql/handler/adminstatement/ExplainHandler.java
package io.jsql.sql.handler.adminstatement;

import com.alibaba.druid.sql.ast.SQLStatement;
import com.alibaba.druid.sql.ast.statement.SQLExplainStatement;
import io.jsql.config.ErrorCode;
import io.jsql.config.Fields;
import io.jsql.mysql.PacketUtil;
import io.jsql.mysql.mysql.FieldPacket;
import io.jsql.sql.OConnection;
import io.jsql.sql.handler.SqlStatementHander;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.util.regex.Pattern;

/**
 * @author jsql
 */
@Component
public class ExplainHandler extends SqlStatementHander{

    private static final Logger logger = LoggerFactory.getLogger(ExplainHandler.class);
    private final static Pattern pattern = Pattern.compile("(?:(\\s*next\\s+value\\s+for\\s*MYCATSEQ_(\\w+))(,|\\)|\\s)*)+", Pattern.CASE_INSENSITIVE);
    //    private static final RouteResultsetNode[] EMPTY_ARRAY = new RouteResultsetNode[0];
    private static final int FIELD_COUNT = 2;
    private static final FieldPacket[] fields = new FieldPacket[FIELD_COUNT];

    static {
        fields[0] = PacketUtil.getField("DATA_NODE",
                Fields.FIELD_TYPE_VAR_STRING);
        fields[1] = PacketUtil.getField("SQL", Fields.FIELD_TYPE_VAR_STRING);
    }

    public static void handle(String stmt, OConnection c, int offset) {
        stmt = stmt.substring(offset).trim();

//        RouteResultset rrs = null;
//        if (rrs == null) {
//            return;
//        }
//
//        ByteBuffer buffer = c.allocate();

        // write header
//        ResultSetHeaderPacket header = PacketUtil.getHeader(FIELD_COUNT);
//        byte packetId = header.packetId;
//        buffer = header.write(buffer, c, true);
//
//        // write fields
//        for (FieldPacket field : fields) {
//            field.packetId = ++packetId;
//            buffer = field.write(buffer, c, true);
//        }
//
//        // write eof
//        EOFPacket eof = new EOFPacket();
//        eof.packetId = ++packetId;
//        buffer = eof.write(buffer, c, true);
//
//        // write rows
//        RouteResultsetNode[] rrsn = rrs.getNodes();
//        for (RouteResultsetNode node : rrsn) {
//            RowDataPacket row = getRow(node, c.getCharset());
//            row.packetId = ++packetId;
//            buffer = row.write(buffer, c, true);
//        }
//
//        // write last eof
//        EOFPacket lastEof = new EOFPacket();
//        lastEof.packetId = ++packetId;
//        buffer = lastEof.write(buffer, c, true);
//
//        // post write
//        c.write(buffer);

    }

//    private static RowDataPacket getRow(RouteResultsetNode node, String charset) {
//        RowDataPacket row = new RowDataPacket(FIELD_COUNT);
//        row.add(StringUtil.encode(node.getName(), charset));
//        row.add(StringUtil.encode(node.getStatement().replaceAll("[\\t\\n\\r]", " "), charset));
//        return row;
//    }

//    private static boolean isMycatSeq(String stmt, SchemaConfig schema) {
//        if (pattern.matcher(stmt).find()) {
//            return true;
//        }
//        SQLStatementParser parser = new MySqlStatementParser(stmt);
//        MySqlInsertStatement statement = (MySqlInsertStatement) parser.parseStatement();
//        String tableName = statement.getTableName().getSimpleName();
//        TableConfig tableConfig = schema.getTables().get(tableName.toUpperCase());
//        if (tableConfig == null) {
//            return false;
//        }
//        if (tableConfig.isAutoIncrement()) {
//            boolean isHasIdInSql = false;
//            String primaryKey = tableConfig.getPrimaryKey();
//            List<SQLExpr> columns = statement.getColumns();
//            for (SQLExpr column : columns) {
//                String columnName = column.toString();
//                if (primaryKey.equalsIgnoreCase(columnName)) {
//                    isHasIdInSql = true;
//                    break;
//                }
//            }
//            if (!isHasIdInSql) {
//                return true;
//            }
//        }
//
//
//        return false;
//    }

    public static void handle(SQLExplainStatement x, OConnection connection) {
        connection.writeErrMessage(ErrorCode.ER_CHECK_NO_SUCH_TABLE, "not soupot");
    }

    @Override
    public Class<? extends SQLStatement> supportSQLstatement() {
        return SQLExplainStatement.class;
    }

    @Override
    protected Object handle(SQLStatement sqlStatement) throws Exception {
        return null;
    }
}

File: src/main/java/io/jsql/sql/handler/adminstatement/ExplainHandler.kt
/*
 * Java-based distributed database like Mysql
 */

package io.jsql.sql.handler.adminstatement

import com.alibaba.druid.sql.ast.SQLStatement
import com.alibaba.druid.sql.ast.statement.SQLExplainStatement
import com.orientechnologies.orient.core.db.document.ODatabaseDocument
import com.orientechnologies.orient.core.record.OElement
import com.orientechnologies.orient.core.record.impl.ODocument
import io.jsql.config.ErrorCode
import io.jsql.config.Fields
import io.jsql.mysql.PacketUtil
import io.jsql.mysql.mysql.FieldPacket
import io.jsql.sql.OConnection
import io.jsql.sql.handler.MyResultSet
import io.jsql.sql.handler.SqlStatementHander
import io.jsql.storage.StorageException
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.util.regex.Pattern

/**
 * 此类用于解析explain select*from table
 * @author jsql
 */
@Component
class ExplainHandler : SqlStatementHander() {

    override fun supportSQLstatement(): Class<out SQLStatement> {
        return SQLExplainStatement::class.java
    }

    @Throws(Exception::class)
    override fun handle0(sqlStatement: SQLStatement, c: OConnection): Any? {
        val explain=sqlStatement as SQLExplainStatement
        val documentTx: ODatabaseDocument
        try {
            documentTx = OConnection.DB_ADMIN!!.getdb(c.schema!!)
        } catch (e: StorageException) {
            e.printStackTrace()
            return e.message!!
        }
        documentTx.activateOnCurrentThread()
        var tablename =explain.toString().split(" ")
        val oClass=documentTx.getClass(tablename[2])
        val element = ODocument()
        var elements=ArrayList<OElement>()

        element.setProperty("id","1")
        element.setProperty("select_type","simple")
        /*
        SIMPLE - 简单的 SELECT （没有使用 UNION 或子查询
        PRIMARY - 最外层的 SELECT
        UNION - 第二层，在SELECT 之后使用了 UNIO
        DEPENDENT UNION - UNION 语句中的第二个 SELECT，依赖于外部子查
        SUBQUERY - 子查询中的第一个 SELEC
        DEPENDENT SUBQUERY - 子查询中的第一个 SUBQUERY 依赖于外部的子查
        DERIVED - 派生表 SELECT（FROM 子句中的子查询
        */
        element.setProperty("table",oClass.toString())
        element.setProperty("partitions","")
        element.setProperty("type","all")//表连接类型
        element.setProperty("possible_keys","NULL")
        element.setProperty("key","NULL")
        element.setProperty("key_len","NULL")
        element.setProperty("ref","NULL")
        element.setProperty("rows",oClass.properties().size)
        element.setProperty("filtered","100.00")
        element.setProperty("extra","NULL")
        elements.add(element)




        return MyResultSet(elements, listOf("id","select_type","table","partitions","type","possible_keys","key","key_len","ref","rows","filtered","extra"))

    }

    companion object {

        private val logger = LoggerFactory.getLogger(ExplainHandler::class.java)
        private val pattern = Pattern.compile("(?:(\\s*next\\s+value\\s+for\\s*MYCATSEQ_(\\w+))(,|\\)|\\s)*)+", Pattern.CASE_INSENSITIVE)
        //    private static final RouteResultsetNode[] EMPTY_ARRAY = new RouteResultsetNode[0];
        private val FIELD_COUNT = 2
        private val fields = arrayOfNulls<FieldPacket>(FIELD_COUNT)

        init {
            fields[0] = PacketUtil.getField("DATA_NODE",
                    Fields.FIELD_TYPE_VAR_STRING)
            fields[1] = PacketUtil.getField("SQL", Fields.FIELD_TYPE_VAR_STRING)
        }

        fun handle(stmt: String, c: OConnection, offset: Int) {
            var stmt = stmt
            stmt = stmt.substring(offset).trim { it <= ' ' }



        fun handle(x: SQLExplainStatement, connection: OConnection) {
            connection.writeErrMessage(ErrorCode.ER_CHECK_NO_SUCH_TABLE, "not soupot")
        }
    }
}}
